src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3152,6 +3153,312 @@ void MacroAssembler::check_klass_subtype(Register sub_klass,
+    z_bfalse(L_failure);
+
+    z_rllg(r_bitmap, r_bitmap, 64-1); // rotate right
+    add2reg(r_array_index, BytesPerWord);
+
+    z_bru(L_loop);
+  }
+
+  { // Degenerate case: more than 64 secondary supers.
+    // FIXME: We could do something smarter here, maybe a vectorized
+    // comparison or a binary search, but is that worth any added
+    // complexity?
+
+    bind(L_huge);
+    repne_scan(r_array_base, r_super_klass, r_array_length, r_result);
+
+    z_bru(L_done); // whatever we got from repne_scan, forward it
+  }
+
+  bind(L_failure);
+  z_lghi(r_result, 1);
+
+  bind(L_done);
+  BLOCK_COMMENT("} lookup_secondary_supers_table_slow_path");
+}
+
+// Make sure that the hashed lookup and a linear scan agree.
+void MacroAssembler::verify_secondary_supers_table(Register r_sub_klass,
+                                                   Register r_super_klass,
+                                                   Register r_result /* expected */,
+                                                   Register r_temp1,
+                                                   Register r_temp2,
+                                                   Register r_temp3) {
+  assert_different_registers(r_sub_klass, r_super_klass, r_result, r_temp1, r_temp2, r_temp3);
+
+  const Register
+    r_array_base   = r_temp1,
+    r_array_length = r_temp2,
+    r_array_index  = r_temp3,
+    r_bitmap       = noreg; // unused
+
+  const Register r_one = Z_R0_scratch;
+  z_lghi(r_one, 1); // for locgr down there, to a load result for failure
+
+  LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS;
+
+  BLOCK_COMMENT("verify_secondary_supers_table {");
+
+#ifdef ASSERT
+  {
+    // r_result should have either 0 or 1 value
+
+    // check for 0
+    z_chi(r_result, 0);
+    asm_assert(bcondNotLow, "r_result should be equal or greater than 0", 33);
+
+    // check for 1
+    z_chi(r_result, 1);
+    asm_assert(bcondNotHigh, "r_result should be equal or less than 1", 33);
+  }
+#endif // ASSERT
+
+  Label L_passed, L_failure;
+
+  // We will consult the secondary-super array.
+  z_lg(r_array_base, Address(r_sub_klass, in_bytes(Klass::secondary_supers_offset())));
+
+  // Load the array length.
+  z_llgf(r_array_length, Address(r_array_base, Array<Klass*>::length_offset_in_bytes()));
+
+  // And adjust the array base to point to the data.
+  z_aghi(r_array_base, Array<Klass*>::base_offset_in_bytes());
+
+  const Register r_linear_result = r_array_index; // reuse
+  z_chi(r_array_length, 0);
+  z_locgr(r_linear_result, r_one, bcondNotHigh); // load failure if array_length <= 0
+  z_brc(bcondNotHigh, L_failure);
+  repne_scan(r_array_base, r_super_klass, r_array_length, r_linear_result);
+  bind(L_failure);
+
+  z_cr(r_result, r_linear_result);
+  z_bre(L_passed);
+
+  lgr_if_needed(Z_ARG1, r_super_klass);
+  lgr_if_needed(Z_ARG2, r_sub_klass);
+  z_lgr(Z_ARG3, r_linear_result);
+  z_lgr(Z_ARG4, r_result);
+  const char* msg = "mismatch";
+  load_const_optimized(Z_ARG5, (address)msg);
```

Comment:
Did you test this? It breaks when you have a register collision (see my assert_different_registers on PPC64). You can test it by removing the `z_bre` above and checking if the arguments are correct.

Reply:
I found one log, from when I was implementing this: 
```
fatal error: mismatch: java.lang.Integer implements java.util.concurrent.Callable: is_subtype_of: 0; linear_search: 0; table_lookup: 1

Context: repne_scan wasn't working properly, table_lookup came up to "1" and repne_scan returned "0" so check failed.
```
 For surety let me just add the asserts as well and see what happens.